README.txt
Description
---------------------

chik (child process killer) is a C++ Windows tool that, when given a command,
executes it by spawning a "cmd" child process and when receiving a TerminateProcess,
also kills all child processes _recursively_. Otherwise the tool tries to act
transparent and behave in the same way as the command given.

This overrides the Windows default behavior of letting child processes live when
their parent is killed, which is usually the Linux behavior. Thus when using this
tool on Windows and a same-named command-forwarding linux script, one can achieve
platform-transparent command execution.


Example
---------------------

As an example take an "external tool" in Eclipse, which spawns a gradle process
to build some java project and gradle spawns multiple child "worker" processes.
When hitting the terminate button in Eclipse, on linux-systems, gradle and all workers
are killed recursively, but in Windows, only gradle, but not the workers are killed,
thus being a major annoyance.
When now using chik and passing the gradle command line to it, a terminate from
Eclipse kills: chik, the gradle process, and the gradle workers.

Compare the process tree:

Eclipse
|-gradle
  |-worker1
  |-worker2

Terminate Results in:

Eclipse
worker1
worker2

With chik:

Eclipse
|-chik
  |-gradle
    |-worker1
    |-worker2

After Terminate:

Eclipse


Implementation
---------------------

chik uses the Windows "Jobs" API (see source) to create a job
and add itself to it. After this, all newly spawned child processes
are automatically part of that same job.

When a parent process in a job is killed, all child processes in that same job
are also killed, thus implementing the desired behavior.

Enjoy!
Thomas

chik/ReleaseNotes.md
Version 1.2
---------------
 - Add restart parameter for watchdog mechanism (contributor: hubugui)

Version 1.1
---------------
Fixing issue #1 https://github.com/thomas-kl/chik/issues/1
 - Now linking statically to C and C++ runtimes.

ChildKiller/ChildKiller.cpp
// License: GPL Version 3.0. See License.txt
// Author: Thomas Klambauer
// Email: Thomas AT Klambauer.info

#include <stdio.h>

#define WIN32_LEAN_AND_MEAN
#define _WIN32_WINNT _WIN32_WINNT_WIN7

#include "WinSDKVer.h"

#include "Windows.h"
#include <string>
#include <iostream>

using namespace std;

void printLastError() {
	wcerr << L"Error: GetLastError= " << GetLastError() << endl;
}

int main(int argc, char* argv[]) {

	if( argc < 2 ) {
		wcout << "ChildKiller Version 1" << endl;
		wcout << " Author: Thomas AT Klambauer.info" << endl;
		wcout << " License: GPL Version 3" << endl;
		wcout << endl;
		wcout << L"ChildKiller [Command] [Arg1] [Arg2] ..." << endl;
		wcout << L"  Windows tool, that executes the given command and arguments via the system" << endl;
		wcout << L"  function, while ensuring that killing the original ChildKiller process will" << endl;
		wcout << L"  also kill all child processes spawned by command." << endl;
		return 0;
	}

	HANDLE jobHandle = CreateJobObject(NULL, NULL);
	if( jobHandle == NULL ) {
		wcerr << L"CreateJobObject failed" << endl;
		printLastError();
		return 1;
	}

	HANDLE thisProcess = GetCurrentProcess();

	if( thisProcess == NULL ) {
		wcerr << L"GetCurrentProcess failed" << endl;
		printLastError();
		return 1;
	}

	JOBOBJECT_BASIC_LIMIT_INFORMATION info = {0};
    info.LimitFlags = JOB_OBJECT_LIMIT_KILL_ON_JOB_CLOSE;

    JOBOBJECT_EXTENDED_LIMIT_INFORMATION extendedInfo;
    extendedInfo.BasicLimitInformation = info;

    int length = sizeof(JOBOBJECT_EXTENDED_LIMIT_INFORMATION);

    if (!SetInformationJobObject(jobHandle, JobObjectExtendedLimitInformation, &extendedInfo, length)) {
		wcerr << L"SetInformationJobObject failed" << endl;
		printLastError();
		return 1;
	}

	if( AssignProcessToJobObject(jobHandle, thisProcess) == 0 ) {
		wcerr << L"AssignProcessToJobObject failed" << endl;
		printLastError();
		return 1;
	}

	if( argc > 1 ) {
		string command;
		// System also only uses cmd /c to execute the given string.
		// Add another cmd with /S to ensure that quote-processing behavior
		// as specified by cmd doc. That behavior is to strip the first quote
		// character and the last one in the string and leave the rest.
		// For that, a leading quote is added here... and one at the end.
		command.append("cmd /S /C \"");
		// Ignore argv[0] (contains this executable's path).
		for( int i = 1; i < argc; i++) {
			// Quote all the params to forward them exactly as we received them.
			command.append("\"");
			command.append(argv[i]);
			command.append("\" ");
		}
		// This quote is also removed by cmd.
		command.append("\"");
		return system(command.c_str());
	}

	return 0;
}



chik/chik.cpp
// License: GPL Version 3.0. See License.txt

#define WIN32_LEAN_AND_MEAN
// As per docs of CreateJobObject function (available starting with Windows XP)
#define _WIN32_WINNT 0x0500 

#include <stdio.h>

#include "WinSDKVer.h"

#include "Windows.h"
#include <string>
#include <iostream>
#include "WinError.h"

using namespace std;

DWORD printError(wstring functionName, DWORD error) {
	wcerr << L"Error from " << functionName << L": GetLastError= " << error << endl;
	return error;
}

DWORD printLastError(wstring functionName) {
	return printError(functionName, GetLastError());
}

int wmain(int argc, wchar_t* argv[]) {

	// 1st arg contains this executable's path. 2nd is first parameter
	if (argc < 2) {
		wcout << L"chik Version 1.2" << endl;
		wcout << L" License: GPL Version 3" << endl;
		wcout << endl;
		wcout << L"chik [-r] <Command> [Arg1] [Arg2] ..." << endl;
		wcout << L"  Windows tool, that executes the given command and arguments via the system" << endl;
		wcout << L"  function, while ensuring that killing the original chik process will" << endl;
		wcout << L"  also kill all child processes spawned by command." << endl;
		wcout << endl;
		wcout << L"   [-r] (Optional) If present, will restart the command after its termination" << endl;
		return 0;
	}

	// Create an unnamed job.
	HANDLE jobHandle = CreateJobObject(NULL, NULL);
	if (jobHandle == NULL) {
		return printLastError(L"CreateJobObject");
	}

	HANDLE thisProcess = GetCurrentProcess();

	if (thisProcess == NULL) {
		return printLastError(L"GetCurrentProcess");
	}

	JOBOBJECT_BASIC_LIMIT_INFORMATION info = { 0 };
	info.LimitFlags = JOB_OBJECT_LIMIT_KILL_ON_JOB_CLOSE;

	JOBOBJECT_EXTENDED_LIMIT_INFORMATION extendedInfo;
	extendedInfo.BasicLimitInformation = info;

	int length = sizeof(JOBOBJECT_EXTENDED_LIMIT_INFORMATION);

	if (!SetInformationJobObject(jobHandle, JobObjectExtendedLimitInformation, &extendedInfo, length)) {
		return printLastError(L"SetInformationJobObject");
	}

	if (AssignProcessToJobObject(jobHandle, thisProcess) == 0) {
		return printLastError(L"AssignProcessToJobObject");
	}

	if (argc > 1) {
		wstring command;
		int i = 0;
		bool restart_child = false;

		/* restart child process */
		if (wcscmp(argv[1], L"-r") == 0) {
			i = 2;
			restart_child = true;
		} else {
			i = 1;
			restart_child = false;
		}

		// Ignore argv[0]
		for (; i < argc; i++) {
			wstring arg(argv[i]);

			// If parameter contains a space, quote the params
			//  to forward them exactly as we received them.
			if (arg.find(L" ") != string::npos) {
				arg = L"\"" + arg + L"\"";
			}
			command.append(arg);
			// Add a space to separate the final args, but no trailing space.
			if (i != argc - 1) {
				command.append(L" ");
			}
		}

		// This will cause windows to spawn a "cmd.exe /c" process
		if (restart_child) {
			while (true) {
				_wsystem(command.c_str());
			}
		}
		else {
			return _wsystem(command.c_str());
		}
		return 0;
	}

	return 0;
}

